app/routers/api/browser.py

```
@@ -50,6 +50,44 @@ def list_directory(
+            # Check if resolved_path is relative to any allowed root
+            for root in allowed_roots:
+                try:
+                    resolved_path.relative_to(root)
+                    allowed = True
+                    break
+                except ValueError:
+                    continue
+
+            if not allowed:
+                logger.warning(
+                    f"Access denied: User {current_user.username} tried to browse restricted path {resolved_path}"
```

Ruff (0.14.14)

[warning] Logging statement uses f-string

(G004)